packages/shared/lib/participation/staking.ts

```diff
@@ -464,3 +487,209 @@ export const hasAccountReachedMinimumAirdrop = (): boolean => {
 
     return overview.assemblyRewards > 0 || overview.shimmerRewards > 0
 }
+
+function isValidPeriodNumber(airdrop: StakingAirdrop, periodNumber: number): boolean {
+    if (!airdrop || periodNumber < 1) return false
+
+    switch (airdrop) {
+        case StakingAirdrop.Assembly:
+            return periodNumber <= LAST_ASSEMBLY_STAKING_PERIOD
+        case StakingAirdrop.Shimmer:
+            return periodNumber <= LAST_SHIMMER_STAKING_PERIOD
+        default:
+            return false
+    }
+}
+
+function getStakingResultUrl(airdrop: StakingAirdrop, periodNumber: number): string {
+    switch (airdrop) {
+        case StakingAirdrop.Assembly:
+            return ASSEMBLY_STAKING_RESULT_URLS[periodNumber - 1] ?? ''
+        case StakingAirdrop.Shimmer:
+            return SHIMMER_STAKING_RESULT_URLS[periodNumber - 1] ?? ''
+        default:
+            return ''
+    }
+}
+
+async function fetchStakingResult(airdrop: StakingAirdrop, periodNumber: number): Promise<StakingPeriodResult> {
+    const stakingResultUrl = getStakingResultUrl(airdrop, periodNumber)
+    const stakingResultResponse = await fetch(stakingResultUrl, getJsonRequestOptions())
+
+    return stakingResultResponse.json()
+}
+
+function getCurrentStakingRewards(): AccountStakingRewards[] {
+    const currentStakingRewards = get(activeProfile)?.stakingRewards ?? []
+    if (currentStakingRewards.length === 0) {
+        return get(get(wallet).accounts).map((account) => ({ accountId: account.id }))
+    } else {
+        return currentStakingRewards
+    }
+}
+
+function getEd25519AddressesOfAccount(account: WalletAccount): string[] {
+    if (!account) return []
+
+    return account.addresses.map((bech32Address) => toHexString(Array.from(Bech32.decode(bech32Address.address).data)))
+}
+
+function getStakingPeriodForAccount(
+    account: WalletAccount,
+    stakingResult: StakingPeriodResult,
+    periodNumber: number
+): StakingPeriod {
+    const ed25519Addresses = getEd25519AddressesOfAccount(account)
+
+    const ed25519AddressesWithRewards = ed25519Addresses
+        .filter((address) => address in stakingResult.rewards)
+        .map((address) => [address, stakingResult.rewards[address]])
+
+    const totalPeriodRewards = ed25519AddressesWithRewards.reduce(
+        (sum: number, current: [string, number]) => sum + current[1],
+        0
+    )
+    const rewards: StakingPeriodRewards = Object.fromEntries(ed25519AddressesWithRewards)
+
+    return {
+        periodNumber,
+        totalPeriodRewards,
+        rewards,
+    }
+}
+
+function getAirdropStakingRewards(
+    currentAccountStakingRewards: AccountStakingRewards,
```

Comment:
Perhaps rename this to cached/previous etc as current is confusing with the current event